File: Modules/Eco.py
import json
import os

class Entity:
    def __init__(self, name):
        self.name = name

class Organism(Entity):
    def __init__(self, name, species, age):
        super().__init__(name)
        self.species = species
        self.age = age
        self.is_alive = True
        self.health = 100
        self.oxygen = 100
        self.temperature_tolerance = {'min': 0, 'max': 100}
        self.disease_resistance = 100
        self.sanity = 100

    def move(self):
        print(f"{self.species} {self.name} is moving.")

    def eat(self):
        print(f"{self.species} {self.name} is eating.")

    def reproduce(self):
        print(f"{self.species} {self.name} is reproducing.")

    def die(self):
        print(f"{self.species} {self.name} has died.")
        self.is_alive = False

    def adjust_oxygen(self, amount):
        self.oxygen += amount
        print(f"{self.species} {self.name} oxygen level adjusted to {self.oxygen}.")

    def adjust_temperature(self, amount):
        self.temperature_tolerance['min'] += amount
        self.temperature_tolerance['max'] += amount
        print(f"{self.species} {self.name} temperature tolerance adjusted to {self.temperature_tolerance['min']}-{self.temperature_tolerance['max']}°C.")

    def adjust_disease_resistance(self, amount):
        self.disease_resistance += amount
        print(f"{self.species} {self.name} disease resistance adjusted to {self.disease_resistance}.")

    def adjust_sanity(self, amount):
        self.sanity += amount
        print(f"{self.species} {self.name} sanity adjusted to {self.sanity}.")

class Animal(Organism):
    def __init__(self, name, species, age, limbs, temperature_preference):
        super().__init__(name, species, age)
        self.limbs = limbs
        self.hunger = 10
        self.strength = 1
        self.gender = 'male' if random.random() < 0.5 else 'female'
        self.temperature_preference = temperature_preference
        self.is_hungry = True

    def attack(self, type):
        damage = 0
        if type == 'bite':
            damage = self.strength * 2
        elif type == 'claw':
            damage = self.strength * 1.5
        elif type == 'tailSwipe':
            damage = self.strength * 1.2
        else:
            damage = self.strength
        print(f"{self.species} {self.name} performs a {type} attack, dealing {damage} damage.")
        return damage

    def defend(self):
        print(f"{self.species} {self.name} is defending.")

    def evade(self):
        print(f"{self.species} {self.name} is evading.")

    def move(self):
        print(f"{self.species} {self.name} is running.")

    def eat(self):
        print(f"{self.species} {self.name} is hunting.")
        self.hunger -= 20
        if self.hunger <= 0:
            self.is_hungry = False

    def lose_limb(self):
        if self.limbs > 0:
            self.limbs -= 1
            print(f"{self.species} {self.name} lost a limb! Remaining limbs: {self.limbs}")
        else:
            self.die()

    def adjust_health(self, amount):
        self.health += amount
        print(f"{self.species} {self.name} health adjusted to {self.health}.")

class Ecosystem:
    def __init__(self):
        self.entities = []
        self.temperature = 25
        self.log = []
        self.day = 0

    def add_entity(self, entity):
        self.entities.append(entity)

    def simulate_day(self):
        print(f"Simulating Day {self.day} in the ecosystem:")
        for entity in self.entities:
            if isinstance(entity, Organism) and entity.is_alive:
                entity.move()
                entity.eat()
                entity.reproduce()
                entity.adjust_oxygen(-5)
                entity.adjust_temperature(1)
                entity.adjust_disease_resistance(-1)
                entity.adjust_sanity(-1)
                if isinstance(entity, Animal):
                    entity.lose_limb()
                    if entity.is_hungry:
                        entity.die()
        self.temperature += random.uniform(-1, 1)
        print(f"Current temperature: {self.temperature:.2f}°C")
        self.log.append(f"Day {self.day}: Temperature changed to {self.temperature:.2f}°C.")
        self.day += 1

    def clean_ecosystem(self):
        self.entities = [entity for entity in self.entities if entity.is_alive]

    def check_ecosystem_health(self):
        alive_organisms = [entity for entity in self.entities if isinstance(entity, Organism) and entity.is_alive]
        total_organisms = len([entity for entity in self.entities if isinstance(entity, Organism)])
        health_percentage = (len(alive_organisms) / total_organisms) * 100
        print(f"Ecosystem health: {health_percentage:.2f}%")

    def display_log(self):
        print("Event Log:")
        for event in self.log:
            print(event)

    def save(self, filename):
        app_dir = '/robjam1990/Psychosis/Gameplay/Planets/Thear/'
        file_path = os.path.join(app_dir, filename)
        data = {
            'entities': [entity.__dict__ for entity in self.entities],
            'temperature': self.temperature,
            'log': self.log,
            'day': self.day
        }
        with open(file_path, 'w') as file:
            json.dump(data, file)
        print(f"Ecosystem saved to {file_path}")

    def load(self, filename):
        app_dir = '/robjam1990/Psychosis/Gameplay/Planets/Thear/'
        file_path = os.path.join(app_dir, filename)
        with open(file_path, 'r') as file:
            saved_state = json.load(file)
            self.entities = [Organism(**entity) if 'limbs' not in entity else Animal(**entity) for entity in saved_state['entities']]
            self.temperature = saved_state['temperature']
            self.log = saved_state['log']
            self.day = saved_state['day']
        print(f"Ecosystem loaded from {file_path}")

# Test the ecosystem save and load functions
ecosystem = Ecosystem()
predators = [Animal("Fox", "Fox", 3, 4, {'min': 20, 'max': 30}), Animal("Wolf", "Wolf", 4, 4, {'min': 15, 'max': 25})]
for predator in predators:
    ecosystem.add_entity(predator)

# Save the ecosystem
ecosystem.save('ecosystem.json')

# Create a new ecosystem and load the saved state
new_ecosystem = Ecosystem()
new_ecosystem.load('ecosystem.json')

# Display event log of the loaded ecosystem
new_ecosystem.display_log()

File: Modules/Eco.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;

public class Entity
{
    public string Name { get; set; }

    public Entity(string name)
    {
        Name = name;
    }
}

public class Organism : Entity
{
    public string Species { get; set; }
    public int Age { get; set; }
    public bool IsAlive { get; set; }
    public int Health { get; set; }
    public int Oxygen { get; set; }
    public TemperatureTemperatureTolerance TemperatureTolerance { get; set; }
    public int DiseaseResistance { get; set; }
    public int Sanity { get; set; }

    public Organism(string name, string species, int age) : base(name)
    {
        Species = species;
        Age = age;
        IsAlive = true;
        Health = 100;
        Oxygen = 100;
        TemperatureTolerance = new TemperatureTemperatureTolerance { Min = 0, Max = 100 };
        DiseaseResistance = 100;
        Sanity = 100;
    }

    public void Move()
    {
        Console.WriteLine($"{Species} {Name} is moving.");
    }

    public void Eat()
    {
        Console.WriteLine($"{Species} {Name} is eating.");
    }

    public void Reproduce()
    {
        Console.WriteLine($"{Species} {Name} is reproducing.");
    }

    public void Die()
    {
        Console.WriteLine($"{Species} {Name} has died.");
        IsAlive = false;
    }

    public void AdjustOxygen(int amount)
    {
        Oxygen += amount;
        Console.WriteLine($"{Species} {Name} oxygen level adjusted to {Oxygen}.");
    }

    public void AdjustTemperature(int amount)
    {
        TemperatureTolerance.Min += amount;
        TemperatureTolerance.Max += amount;
        Console.WriteLine($"{Species} {Name} temperature tolerance adjusted to {TemperatureTolerance.Min}-{TemperatureTolerance.Max}°C.");
    }

    public void AdjustDiseaseResistance(int amount)
    {
        DiseaseResistance += amount;
        Console.WriteLine($"{Species} {Name} disease resistance adjusted to {DiseaseResistance}.");
    }

    public void AdjustSanity(int amount)
    {
        Sanity += amount;
        Console.WriteLine($"{Species} {Name} sanity adjusted to {Sanity}.");
    }
}

public class TemperatureTemperatureTolerance
{
    public int Min { get; set; }
    public int Max { get; set; }
}

public class Animal : Organism
{
    public int Limbs { get; set; }
    public int Hunger { get; set; }
    public int Strength { get; set; }
    public string Gender { get; set; }
    public TemperatureTemperatureTolerance TemperaturePreference { get; set; }
    public bool IsHungry { get; set; }

    public Animal(string name, string species, int age, int limbs, TemperatureTemperatureTolerance temperaturePreference) : base(name, species, age)
    {
        Limbs = limbs;
        Hunger = 10;
        Strength = 1;
        Gender = new Random().NextDouble() < 0.5 ? "male" : "female";
        TemperaturePreference = temperaturePreference;
        IsHungry = true;
    }

    public int Attack(string type)
    {
        int damage;
        switch (type)
        {
            case "bite":
                damage = Strength * 2;
                break;
            case "claw":
                damage = Strength * 1.5;
                break;
            case "tailSwipe":
                damage = Strength * 1.2;
                break;
            default:
                damage = Strength;
                break;
        }
        Console.WriteLine($"{Species} {Name} performs a {type} attack, dealing {damage} damage.");
        return damage;
    }

    public void Defend()
    {
        Console.WriteLine($"{Species} {Name} is defending.");
    }

    public void Evade()
    {
        Console.WriteLine($"{Species} {Name} is evading.");
    }

    public new void Move()
    {
        Console.WriteLine($"{Species} {Name} is running.");
    }

    public new void Eat()
    {
        Console.WriteLine($"{Species} {Name} is hunting.");
        Hunger -= 20;
        if (Hunger <= 0)
        {
            IsHungry = false;
        }
    }

    public void LoseLimb()
    {
        if (Limbs > 0)
        {
            Limbs--;
            Console.WriteLine($"{Species} {Name} lost a limb! Remaining limbs: {Limbs}");
        }
        else
        {
            Die();
        }
    }

    public void AdjustHealth(int amount)
    {
        Health += amount;
        Console.WriteLine($"{Species} {Name} health adjusted to {Health}.");
    }
}

public class Ecosystem
{
    public List<Entity> Entities { get; set; }
    public int Temperature { get; set; }
    public List<string> Log { get; set; }
    public int Day { get; set; }

    public Ecosystem()
    {
        Entities = new List<Entity>();
        Temperature = 25;
        Log = new List<string>();
        Day = 0;
    }

    public void AddEntity(Entity entity)
    {
        Entities.Add(entity);
    }

    public void SimulateDay()
    {
        Console.WriteLine($"Simulating Day {Day} in the ecosystem:");
        foreach (var entity in Entities)
        {
            if (entity is Organism organism && organism.IsAlive)
            {
                organism.Move();
                organism.Eat();
                organism.Reproduce();
                organism.AdjustOxygen(-5);
                organism.AdjustTemperature(1);
                organism.AdjustDiseaseResistance(-1);
                organism.AdjustSanity(-1);
                if (entity is Animal animal)
                {
                    animal.LoseLimb();
                    if (animal.IsHungry)
                    {
                        animal.Die();
                    }
                }
            }
        }
        Temperature += new Random().Next(-1, 2);
        Console.WriteLine($"Current temperature: {Temperature}°C");
        Log.Add($"Day {Day}: Temperature changed to {Temperature}°C.");
        Day++;
    }

    public void CleanEcosystem()
    {
        Entities.RemoveAll(entity => entity is Organism organism && !organism.IsAlive);
    }

    public void CheckEcosystemHealth()
    {
        var aliveOrganisms = Entities.FindAll(entity => entity is Organism organism && organism.IsAlive);
        var totalOrganisms = Entities.FindAll(entity => entity is Organism).Count;
        var healthPercentage = (double)aliveOrganisms.Count / totalOrganisms * 100;
        Console.WriteLine($"Ecosystem health: {healthPercentage:F2}%");
    }

    public void DisplayLog()
    {
        Console.WriteLine("Event Log:");
        foreach (var eventLog in Log)
        {
            Console.WriteLine(eventLog);
        }
    }

    public void Save(string filename)
    {
        var appDir = "/robjam1990/Psychosis/Gameplay/Planets/Thear/";
        var filePath = Path.Combine(appDir, filename);
        var data = JsonConvert.SerializeObject(new
        {
            entities = Entities,
            temperature = Temperature,
            log = Log,
            day = Day
        });
        File.WriteAllText(filePath, data);
        Console.WriteLine($"Ecosystem saved to {filePath}");
    }

    public void Load(string filename)
    {
        var appDir = "/robjam1990/Psychosis/Gameplay/Planets/Thear/";
        var filePath = Path.Combine(appDir, filename);
        var data = File.ReadAllText(filePath);
        var savedState = JsonConvert.DeserializeObject<EcosystemSaveState>(data);
        Entities = savedState.Entities;
        Temperature = savedState.Temperature;
        Log = savedState.Log;
        Day = savedState.Day;
        Console.WriteLine($"Ecosystem loaded from {filePath}");
    }
}

public class EcosystemSaveState
{
    public List<Entity> Entities { get; set; }
    public int Temperature { get; set; }
    public List<string> Log { get; set; }
    public int Day { get; set; }
}

class Program
{
    static void Main(string[] args)
    {
        // Test the ecosystem save and load functions
        var ecosystem = new Ecosystem();
        var predators = new List<Animal>
        {
            new Animal("Fox", "Fox", 3, 4, new TemperatureTemperatureTolerance { Min = 20, Max = 30 }),
            new Animal("Wolf", "Wolf", 4, 4, new TemperatureTemperatureTolerance { Min = 15, Max = 25 })
        };
        foreach (var predator in predators)
        {
            ecosystem.AddEntity(predator);
        }

        // Save the ecosystem
        ecosystem.Save("ecosystem.json");

        // Create a new ecosystem and load the saved state
        var newEcosystem = new Ecosystem();
        newEcosystem.Load("ecosystem.json");

        // Display event log of the loaded ecosystem
        newEcosystem.DisplayLog();
    }
}
